ui/raidboss/emulator/ui/Tooltip.ts

```diff
@@ -0,0 +1,167 @@
+const hideEvents = [
+  'mouseleave',
+  'blur',
+] as const;
+
+const validDirections = [
+  'top',
+  'right',
+  'bottom',
+  'left',
+] as const;
+type ValidDirection = typeof validDirections[number];
+
+abstract class HTMLTemplateElement extends HTMLElement {
+  abstract get content(): DocumentFragment;
+}
+
+const showEvents = [
+  'mouseenter',
+  'focus',
+] as const;
+
+type TemplatesType = {[Property in ValidDirection]: HTMLTemplateElement};
+
+const toPx = (px: number): string => `${px}px`;
+
+export default class Tooltip {
+  private offset = {
+    x: 0,
+    y: 0,
+  };
+  private target: HTMLElement;
+  private direction: ValidDirection;
+  private tooltip: HTMLElement;
+  private inner: HTMLElement;
+  private arrow: HTMLElement;
+
+  private static templates: TemplatesType;
+
+  constructor(
+      targetRef: string | HTMLElement,
+      direction: ValidDirection,
+      text: string,
+      autoShow = true,
+      autoHide = true) {
+    Tooltip.initializeTemplates();
+
+    let target: HTMLElement | null;
+
+    if (typeof targetRef === 'string')
+      target = document.querySelector<HTMLElement>(targetRef);
+    else
+      target = targetRef;
+
+    if (!target) {
+      const msg = 'Invalid selector or element passed to Tooltip';
+      console.error(msg);
+      throw new Error(msg);
+    }
+
+    this.target = target;
+    this.direction = direction;
+    this.tooltip = Tooltip.cloneTemplate(direction);
+    const innerElem = this.tooltip.querySelector<HTMLElement>('.tooltip-inner');
+    if (!innerElem) {
```

Comment:
This is just a type assertion, right? `querySelector` isn't doing anything fancy with the parameter other than casting? If so, then maybe:

```suggestion
    const innerElem = this.tooltip.querySelector('.tooltip-inner');
    if (!innerElem || !(innerElem instanceof HTMLElement)) {
```